File: nim_core/ios/Classes/SafeMethodChannel.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation

class SafeMethodChannel {
  private var methodChannel: FlutterMethodChannel?
  private var mainQueue = DispatchQueue.main

  init(_ name: String, _ messenger: FlutterBinaryMessenger) {
    methodChannel = FlutterMethodChannel(name: name, binaryMessenger: messenger)
  }

  func getChannel() -> FlutterMethodChannel? {
    methodChannel
  }

  func invokeMethod(_ method: String, _ arguments: Any?) {
    weak var weakSelf = self
    performOnMainthod {
      weakSelf?.methodChannel?.invokeMethod(method, arguments: arguments)
    }
  }

  func invokeMethod(_ method: String, _ arguments: Any?, result: @escaping FlutterResult) {
    weak var weakSelf = self
    performOnMainthod {
      weakSelf?.methodChannel?.invokeMethod(method, arguments: arguments, result: result)
    }
  }

  private func performOnMainthod(_ block: @escaping () -> Void) {
    if Thread.current.isMainThread {
      block()
    } else {
      DispatchQueue.main.async {
        block()
      }
    }
  }
}

File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/SafeMethodChannel.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class SafeMethodChannel(messenger: BinaryMessenger, name: String) {

    private val methodChannel: MethodChannel = MethodChannel(messenger, name)
    private val handler: Handler = Handler(Looper.getMainLooper())

    fun invokeMethod(method: String, arguments: Any?, callback: MethodChannel.Result? = null) {
        runOnMainThread { methodChannel.invokeMethod(method, arguments, callback) }
    }

    fun setMethodCallHandler(handler: MethodCallHandler?) {
        runOnMainThread { methodChannel.setMethodCallHandler(handler) }
    }

    private fun runOnMainThread(runnable: () -> Unit) {
        if (Looper.getMainLooper() == Looper.myLooper()) {
            runnable()
        } else {
            handler.post(runnable)
        }
    }
}
